End-to-end impl for fp16 backward weight convolution using atomic add.

whchung opened the pull request:
Rebased with the tip.

Using atomic adds under the following conditions:
- backward weight convolution
- fp16
- XDLOPS
- do not use padding along Gemm M/N/K dimension

If all conditions are met, an extra workspace is required to hold the result in fp32 data type. An `Miir` interface is added to let clients query how large the workspace should be.

`miopen-gen` is modified to automatically change the kernel signature when such an extra workspace is needed.

Utility functions inside `Conv2dGenerator` are modified to compute if a config requires a workspace. Notice there are hard-coded tuning paremeters in the current implementation which is very not ideal.

Added test cases to check various configs on whether a workspace is needed.

Empirical performance tests show 3.5X performance improvements on supported resnet50 fp16 backward weight convolution configs.

whchung committed:
End-to-end impl for fp16 backward weight convolution using atomic add.

Rebased with the tip.

jerryyin commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1318,7 +1314,7 @@ struct GridwiseGemmV2RewritePattern
 
     // Obtain data type.
-    auto elementType = op.c().getType().cast<MemRefType>().getElementType();
+    auto elementType = op.b().getType().cast<MemRefType>().getElementType();
```
I happen to need to change this as well because the kernel return arg C type is different with A or B matrix. However, how does the fp16 not using atomic add being able to work before (I'm assuming C type is fp32, so the old code sort of give us the element type of fp32?)

whchung replied:
on the non-atomic add version, because each thread stores to a different location on the global memref, we do an in-place fp32->fp16 truncation before storing it.

jerryyin commented on mlir/tools/mlir-miopen-lib/mlir-miopen-lib-test.cpp:
```diff
@@ -43,6 +43,8 @@ int main(int argc, char **argv) {
     std::cout << "ExecutionDims - globalSize=" << globalSize
               << ", localSize=" << localSize << std::endl;
 
+  } else if (option.getValue() == "workspace") {
+    std::cout << "Workspace=" << miirGetWorkspaceSize(handle) << std::endl;
```
Mind adding at least 1 unit test on the library test side with workspace size no equal to zero?

whchung replied:
We do have a test to check non-zero workspace size. Please refer to `workspace.mlir` in this PR.

jerryyin commented on mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:
```diff
@@ -318,6 +318,83 @@ Type Conv2dGenerator::getDataType(OpBuilder &builder) const {
+    if ((config.operation.getValue() == miopen::ConvOpType::BwdWeight) &&
+        config.xdlops && (dataType == builder.getF16Type())) {
+
+      // Logic to check if we need to pad along GemmM/N/K dimension for backward
+      // weight convolution.
+      // FIXME. Merge with logic used in the tuning process.
```
Why do you need to execute a portion of the tuning logic beforehand? I think you only need to check if all preconditions are met. Then if it ended up to be a config that can't satisfy the below, you can let tuning step to reject it naturally.

whchung replied:
I'm open to ideas here. This part is one of the worst place I think in this PR.

This is one of the major dilemma I faced earlier this week solving the unit test failures.

One of the precondition: NOT padding along Gemm M/N/K requires understanding what tuning parameters (M/N/K per block) be picked. 

We can understand the Gemm M/N/K size from the convolution configuration, but in `Conv2dGenerator` we don't yet the what the tuning parameters would be.

And structures inside `include/Dialect/MIOpen/Tuning` and `include/Dialect/MIOpen/Generator` are not compatible. So I had to manually extract logic here.

jerryyin replied:
> One of the precondition: NOT padding along Gemm M/N/K requires understanding what tuning parameters (M/N/K per block) be picked.

Just to understand: When this condition does not meet, we will fall back to the conv2d kernel that does not require a workspace?

Another question: If the answer is yes for the previous question, then does the tuning parameter between the kernel with/out the workspace work between one another?

whchung replied:
If any of the preconditions is not met:
- backward weight convolution
- fp16
- XDLOPS
- not padding along Gemm M/N/K dimension

we will fallback to the conv2d kernel that does not require a workspace.

The tuning parameters with or without workspace can be used interchangeably.

krzysz00 replied:
One question: what breaks when there's gemmM/N/K padding?

Also, re duplicate tuning logic - I think what needs to happen there is to refactor that whole pile of code since we're poking around in there: make some sort of `ConvDefinition` struct that has all the relevant info and then both the generator and tuning code will operate on (or have a) `ConvDefinition` - we can also make `ConvDefinition` convertible to `ConvContext` for serialization purposes, but still.

That'd also allow us to factor out the gemm size computations that have been copied everywhere in `ConvToGemm`

jerryyin replied:
> The tuning parameters with or without workspace can be used interchangeably.

I know that when MIOpen is facing this sort of situation, they will expose both ops. Such that:
 - Both ops are candidate. 
    - One faster op is faster but requires more memory
    - Another slower op is slower but requires less memory
 - The compromise between speed and memory is something that (MIOpen) client can choose in their specific scenarios. 
 - At the tuning time, MIOpen will choose tuning parameter independently for individual ops

I think the question goes back to whether you/we are okay with always choosing the absolutely most high performance kernel over any other constraint? If yes, then this chunk of precondition decision would be totally necessary and there's no other way around a decision like this in conv2d generator.

whchung replied:
> One question: what breaks when there's gemmM/N/K padding?

In `miopen-dialect`, if we deliberately allow padding cases using atomic add kernel, the result would be 0 everywhere on both fp32/fp16 cases. The root cause is not known yet.


> Also, re duplicate tuning logic - I think what needs to happen there is to refactor that whole pile of code since we're poking around in there: make some sort of `ConvDefinition` struct that has all the relevant info and then both the generator and tuning code will operate on (or have a) `ConvDefinition` - we can also make `ConvDefinition` convertible to `ConvContext` for serialization purposes, but still.
> 
> That'd also allow us to factor out the gemm size computations that have been copied everywhere in `ConvToGemm`

Agree an overhaul here is necessary, I see somewhat duplicated structures like `Conv2dGenerator::Config` in the generator directory, and `ConvContext` in the tuning domain.

krzysz00 replied:
And the advantage to the overhaul is that we could memoize the gemm size computation and tuning selection, instead of running that valid tuning parameter section loop a bunch of times.

So yeah, I'd rather this not go in without an overhaul to get rid of this copy-paste.

And can we re-test the padding thing, just to see if something changed in the meantime?

whchung replied:
Let me do an assessment on the scope of overhaul so we can prioritize it. I think an overhaul is necessary, but whether it blocks the PR and subsequent ones (MIOpen solver change, use GPU to do fp32->fp16 zeroing / conversion) is up for discussion.

krzysz00 replied:
The objection I have is that the temporary copy-paste may become much less temporary because new features will get prioritized over fixing this

whchung replied:
That's why we use tickets to track the laundry list of known todo items.

whchung replied:
https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/439 to keep track of the new requirements identified.

whchung replied:
As an initial step I'm thinking to abstract `calculatePaddingKernelSize` used in several places (`-miopen-affix-params`, various kernel generators, and here) into a utility function first.

This should reduce the scope of additional codes be impacted in the PR, while paving way for a more complete overhaul in #439.

whchung replied:
Meanwhile, supplying a vector of heuristic tuning parameters here isn't entirely correct. But works for now.

We supply a vector of tuning parameters based on heuristics in `-miopen-affix-params` out of necessity. If a client application doesn't specify one, we must come up one for later stages.

On the other hand, at `conv2dGenerator` stage, which tuning parameter would be picked isn't known beforehand, nor do we have an interface for client applications to pass in the specification of the tuning search space so the generator could try the entire space.

So basically logic here is mostly like:
- Let's assume the client won't set a tuning parameter, so we just steal the vector we'll try in the next stage `-miopen-affix-params` and run the checks.

This logic above is not ideal, but work most of the time, and work for all of the e2e tests we do right now, because for cases which requires padding kernels, one of Gemm M / N / K would either be a odd number, or not be divided by 4.

whchung replied:
So the dilemma we face here is:
1) `conv2dGenerator` is not a tuner (ex: solver within MIOpen), it possess zero knowledge of the tuning search space.
2) `conv2dGenerator` does not know the internals of the tuning process (M/N/KPerBlock, M/NPerWave, KPACK, etc).

But, to determine whether padding kernel is used requires the knowledge of the tuning parameters possibly be used for a particular convolution configuration.

As a result, in the current form of the PR, since `conv2dGenerator` doesn't know the tuning search space (1 above), it has to be amended to possess part of 2. Therefore we have duplicated logic.

jerryyin replied:
@whchung Actually inspired by what you are doing here. Wouldn't it be possible to derive the tuning parameter in constructing the conv op:
 - In constructing the conv op, derive tuning parameters before a conv op is constructed
   - If succeed, the constructed conv op would have tuning parameters, so there's no need to do it as a lowering step
   - If failed, we'd return a empty handle, allowing the failure to happen even earlier
 - It sounds like if all tuning parameters validation are brought earlier, we can (use the info to) decide what kind of conv op to construct

whchung replied:
To remedy duplicated logic, I'd like create an oracle function to judge whether a padding kernel is needed, given the GemmM/N/K size, and the direction of the convolution.

The oracle function would possess the vector of initial tuning parameters, just like what's in use in `GridwiseGemmParams`.

On the other hand, I will NOT consolidate somewhat duplicated structures between generator and tuning modules: `Conv2dGenerator::Config` / `ConvContext` in this PR. Such refactoring looks completely beyond the scope of the this PR and is better handled as a separate ticket.

whchung replied:
@jerryyin You are talking about eliminating `-miopen-affix-params` as an independent pass and run it inside `conv2dGenerator`? I'm not against the idea but it still seems to be beyond the scope of the PR. I'll log it in https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/439

krzysz00 replied:
I agree with @jerryyin's proposal, more or less (One interesting thing to consider is whether a convOp can hold the set of valid configurations - or if it could generate it - so that  tuning loops can be told what the valid parameters are for a particular configuration)

whchung replied:
Initial version of refactoring is done. `calculatePaddingKernelSize` is made as a function instead a lambda.

jerryyin commented on mlir/lib/Dialect/MIOpen/Transforms/ConvToGemmLowering.cpp:
```diff
@@ -291,8 +301,9 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
     TransformMapAttr addKBlockTransformAttr = addKBlockTransform.get();
-    Value withKBlock =
-        b.create<TransformOp>(loc, op.filter(), addKBlockTransformAttr);
+    Value filterTensorInUse = (hasWorkspace) ? op.workspace()[0] : op.filter();
+    Value withKBlock = b.create<miopen::TransformOp>(loc, filterTensorInUse,
+                                                     addKBlockTransformAttr);
```
Out of curiosity, is this the key logic for using the workspace? I'm asking because I don't see much relevant logic below gridwise lowering.

whchung replied:
Yes. This is where we specify the true output in a backward weight convolution when we use atomic add.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -67,7 +67,8 @@ def MIOpen_Conv2DBwdWeightOp :
                    MemRefRankOf<[F32, F16, BF16], [5]>:$input,
-                   MemRefRankOf<[F32, F16, BF16], [5]>:$output)> {
+                   MemRefRankOf<[F32, F16, BF16], [5]>:$output,
+                   Variadic<MemRefRankOf<[F32], [5]>>:$workspace)> {
```
To be honest, I almost feel like this should be a different high level op.

whchung replied:
I don't agree. Adding one extra top-level op would incur additional overheads in tools such as `Conv2dGenerator` and `miopen-gen`.

I will try make this `Variadic` be `Optional` on the other hand.

krzysz00 replied:
Yeah, I vote for Optional here too

whchung commented:
@jerryyin CI failing at `Test selected MIOpen configs` because MIOpen change to allocate workspace has not been submitted. On my local working MIOpen branch, `mlir-workspace`, the implementation is ready. How should we solve this chicken-and-egg problem?

jerryyin commented:
@whchung In this same PR, make a change in jenkinsfile and swap the miopen develop branch from our CI with your branch. Then after your MIOpen PR is merged into MIOpen, submit another PR to swap that back to develop branch.

whchung commented on mlir/test/mlir-miopen-lib/workspace.mlir:
```diff
@@ -0,0 +1,45 @@
+// RUN: mlir-miopen-lib-test --args " --x2 0 --operation conv2d_bwd_weight --arch amdgcn-amd-amdhsa:gfx908:sramecc+:xnack- --num_cu 120 --in_type fp32 --fil_type fp32 --out_type fp32 --fil_layout GNCHW --in_layout NGCHW --out_layout NGCHW --batchsize 64 --in_channels 1024 --out_channels 1024 --in_h 14 --in_w 14 --out_h 14 --out_w 14 --fil_h 1 --fil_w 1 --dilation_h 1 --dilation_w 1 --conv_stride_ ...
+// RUN: mlir-miopen-lib-test --args " --x2 1 --operation conv2d_bwd_weight --arch amdgcn-amd-amdhsa:gfx908:sramecc+:xnack- --num_cu 120 --in_type fp32 --fil_type fp32 --out_type fp32 --fil_layout GNCHW --in_layout NGCHW --out_layout NGCHW --batchsize 64 --in_channels 1024 --out_channels 1024 --in_h 14 --in_w 14 --out_h 14 --out_w 14 --fil_h 1 --fil_w 1 --dilation_h 1 --dilation_w 1 --conv_stride_ ...
+// fp16+XDLOPS+backward weight. This would require a workspace.
+// RUN: mlir-miopen-lib-test --args " --x2 1 --operation conv2d_bwd_weight --arch amdgcn-amd-amdhsa:gfx908:sramecc+:xnack- --num_cu 120 --in_type fp16 --fil_type fp16 --out_type fp16 --fil_layout GNCHW --in_layout NGCHW --out_layout NGCHW --batchsize 64 --in_channels 1024 --out_channels 1024 --in_h 14 --in_w 14 --out_h 14 --out_w 14 --fil_h 1 --fil_w 1 --dilation_h 1 --dilation_w 1 --conv_stride_ ...
```
@jerryyin This is the config where we explicitly require a workspace, as it fulfills all the criteria:

- fp16
- XDLOPS
- backward weight
- not padding along Gemm M/N/K

jerryyin replied:
Thanks, I somehow thought this is a test through `miopen-gen`. Looks good.

whchung replied:
`miopen-gen` tests don't require `Miir` interface so I didn't add test cases there.

whchung committed:
Temporarily change MIOpen branch from develop to mlir-workspace.

mlir-workspace contains logic to support MLIR xdlops wrw solver to use
a workspace.

whchung committed:
Switch from Variadic to Optional for workspace in conv2d_bwd_weight op.

whchung commented:
jenkins: retest this please.

whchung commented:
Observing something new on CI logs in `Test selected MIOpen configs` stage:
```
"hipErrorNoBinaryForGpu: Unable to find code object for all current devices!"
```
I assume it's a node-specific issue. The very same test command run fine on -112.

krzysz00 commented:
@whchung Yeah, node-specific issue. Means the node needs a GPU reset

Also, instead of "retest this please", you can go into Jenkins and call for the project to be built again

whchung commented:
jenkins: retest this please.

whchung commented:
On CI we already have 2 passing logs but it's not reflected here so I'm triggering once again via jenkins retest command.

jerryyin reviewed (approved):
LGTM. Please keep an eye on weekly (or run through all tuning to make sure things won't break)

whchung committed:
Initial commit to consolidate all call sites of calculatePaddingKernelSize.

whchung commented:
Applying `skip-ci` label as quite a few refactoring commits are coming in.

whchung committed:
Decouple calculatePaddingKernelSize from ConvContext.

whchung committed:
Adopt calculatePaddingKernelSize in conv2dGenerator.

whchung commented:
Implemented the first version of the oracle function `calculatePaddingKernelSize`, the interface is not ideal and can be further amended. I don't quite understand the rationale of the parameter `isOriginalKernelSupport`. It seems to be one of the nuisance caused by backward data kernel.

Remove `skip-ci` and let the CI do some initial checks first. The PR is getting bigger and I'd like to verify it before moving further.

ltqin commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -778,7 +778,7 @@ class PopulateParams : public PopulateParamsBase {
 
   llvm::SmallVector<InitParamsNonXDL, 8>
-  getTuningParameters(ConvolutionContext &ctx) {
+  getTuningParameters(miopen::ConvOpType dir, mlir::Type dataType) {
```
Two parameters are useless?

ltqin commented on mlir/lib/Dialect/MIOpen/Transforms/ConvToGemmLowering.cpp:
```diff
@@ -195,10 +195,20 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
+  bool hasWorkspace =
+      (filterType.getElementType() == b.getF16Type() && isXdlops);
+  if (hasWorkspace) {
+    assert(op.workspace());
```
assert() just run in debug, is there right?

krzysz00 replied:
I think it runs always unless explicitly disabled. However, can we stick an error message on the assert? `assert(op.workspace() && "Op has no workspace")`

whchung replied:
Will do.

krzysz00 commented:
@whchung I think you need to also edit `Jenkinsfile.downstream` to make CI pass

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/ConvToGemmLowering.cpp:
```diff
@@ -1172,56 +1148,32 @@ template <typename T> struct Conv2DRewritePattern : public OpRewritePattern<T> {
-      calculatePaddingKernelSize(populateParams);
+      std::tie(isOriginalKernelSupport, needExtraPad, gemmMExtra, gemmNExtra,
+               gemmKExtra) =
+          calculatePaddingKernelSize(gemmMSize, gemmNSize, gemmKSize,
```
Thanks for pulling these out

krzysz00 commented:
(Update, what you might need to do in Jenkinsfile.downstream is the same fix we have in the main Jenknisfile where you stop the MIOpen build system from building a copy of MLIR with `sed`)

whchung committed:
Modify Jenkinsfile.downstream per review comment.

whchung committed:
Add assert error message per review comment.

whchung commented:
Studying failures on the non-XDLOPS path.
http://ml-ci.amd.com:21096/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-512/20/pipeline/64

MIOpen `mlir-workspace` branch needs to be amended so workspace is only allocated when truly needed. The current implementation merely assumes as long as the following preconditions are met, a workspace is allocated and used:
- fp16
- XDLOPS
- backward weight

It forgot to take padding kernel into consideration. Need to understand when the invoker gets invoked by a solver, does the solver already understands that a workspace is not needed via the amended `Miir` interface.

krzysz00 commented:
@whchung I don't think that's the problem - the failure is because the non-xdlops path compiles the `miopen-dialect` branch as a dependency instead of using the static library built during CI. I now have a PR that changes this so the non-xdlops path behaves like the xdlops one.

whchung commented:
> @whchung I don't think that's the problem - the failure is because the non-xdlops path compiles the `miopen-dialect` branch as a dependency instead of using the static library built during CI. I now have a PR that changes this so the non-xdlops path behaves like the xdlops one.

Thanks. Let me take a look at that PR. On the other hand, existing logic I put together in MIOpen `mlir-workspace` branch is indeed flawed, and I'm working on fixing it.

whchung committed:
Merge branch 'miopen-dialect' into fp16-wrw-optimization-rebase

whchung commented:
Fixed MIOpen `mlir-workspace` branch. Give this PR another go.
